### PR TITLE
Stop attaching test reruns to final test report entries

### PR DESCRIPTION
**Pull request by gnikonorov**
Stop attaching the rerun entries created by [pytest-rerunfailures](https://github.com/pytest-dev/pytest-rerunfailures) to  reports.

When we introduced our post processing step we incorrectly lumped in rerunfailures reruns ( see https://github.com/pytest-dev/pytest-html/pull/271 ), which caused this issue. To fix it, we now treat each rerun as its own test which is what the original behavior was. I confirmed that behavior is the same as v2.1.1 now.

I'd also like to make a release after this merges, since this is a bugfix for the `v3.0.0` release if no one objects

Closes https://github.com/pytest-dev/pytest-html/issues/374

### CODE REVIEW

**Comment by ssbarnea** [comment]
IMHO retries should be visible, they may include essential debugging information, in fact we should also make them visible in the UI with some other color than green.

**Comment by BeyondEvil** [comment]
> IMHO retries should be visible, they may include essential debugging information, in fact we should also make them visible in the UI with some other color than green.

I agree. But I think it might be worth doing a bugfix release, and then adding the functionality you mention in a separate release.

**Comment by gnikonorov** [comment]
@ssbarnea the retries are visible. The issue is that without this PR we take all the retry data and add it to the original run. With this PR the data for each retry row is not added to the original run. Does that make sense?

For example. Here is what a report off of v3.0.0 looks like:
Screen Shot 2020-11-29 at 10 08 13 AM

Here is the same test run used to generate the above report but off of this PR instead:
Screen Shot 2020-11-29 at 10 08 49 AM


You can see that the retry rows still exist and have the proper data. We aren't, however, adding everything into the final pass/fail run of the retry as we are off of head though which is a bug. This didn't happen in 2.1.1 and shouldn't happen IMO.

**Comment by gnikonorov** [comment]
> > IMHO retries should be visible, they may include essential debugging information, in fact we should also make them visible in the UI with some other color than green.
> 
> I agree. But I think it might be worth doing a bugfix release, and then adding the functionality you mention in a separate release.

@BeyondEvil, that functionality already exists, please see the above comment

**Comment by BeyondEvil** [comment]
> @BeyondEvil, that functionality already exists, please see the above comment

Yepp, you're right! 👍

**Review by BeyondEvil — COMMENTED** [review]
Why is the time/duration stuff in this PR?

**Comment by gnikonorov** [comment]
> Why is the time/duration stuff in this PR?

I moved time formatting to the `TestResult` class since it makes no sense to format on the report as we merge reports for test results. Formatting should be done only once when we compute the test result to show to the user. The way we were doing it was that we formatted once per report, and the later merged reports which is wrong as we first need to accumulate all the report times and only then format them for display

**Comment by gnikonorov** [comment]
> > Why is the time/duration stuff in this PR?
> 
> I moved time formatting to the `TestResult` class since it makes no sense to format on the report as we merge reports for test results. Formatting should be done only once when we compute the test result to show to the user. The way we were doing it was that we formatted once per report, and the later merged reports which is wrong as we first need to accumulate all the report times and only then format them for display

And seeing that we needed to touch the logic anyway as part of this change ( line 718 in the original version of plugin.py ), I figured we may as well move it to the proper place instead of moving it but keep a known bug that we'd have to fix for an unreleased feature later on

**Comment by BeyondEvil** [comment]
> > > Why is the time/duration stuff in this PR?
> > 
> > 
> > I moved time formatting to the `TestResult` class since it makes no sense to format on the report as we merge reports for test results. Formatting should be done only once when we compute the test result to show to the user. The way we were doing it was that we formatted once per report, and the later merged reports which is wrong as we first need to accumulate all the report times and only then format them for display
> 
> And seeing that we needed to touch the logic anyway as part of this change ( line 718 in the original version of plugin.py ), I figured we may as well move it to the proper place instead of moving it but keeping a known bug that we'd have to fix for an unreleased feature later on

Would it makes sense to have that as a separate PR? Since it's (sort of) unrelated to reruns? 🤔

**Comment by gnikonorov** [comment]
> > > > Why is the time/duration stuff in this PR?
> > > 
> > > 
> > > I moved time formatting to the `TestResult` class since it makes no sense to format on the report as we merge reports for test results. Formatting should be done only once when we compute the test result to show to the user. The way we were doing it was that we formatted once per report, and the later merged reports which is wrong as we first need to accumulate all the report times and only then format them for display
> > 
> > 
> > And seeing that we needed to touch the logic anyway as part of this change ( line 718 in the original version of plugin.py ), I figured we may as well move it to the proper place instead of moving it but keeping a known bug that we'd have to fix for an unreleased feature later on
> 
> Would it makes sense to have that as a separate PR? Since it's (sort of) unrelated to reruns? 🤔

We could, but I think that would make this change dirty since it would look like this:
```
                 if test_report.outcome == "rerun":
                     # reruns are separate test runs for all intensive purposes
+                    test_report.duration = getattr(test_report, "duration", 0.0)
+                    test_report.formatted_duration = self._format_duration(test_report)
                     self.append_rerun(test_report)
```

And we would also be doing it immediately below on https://github.com/pytest-dev/pytest-html/blob/master/src/pytest_html/plugin.py#L717 which is confusing to read.

I'd argue this is related as this PR is really fixing all postprocessing related to dynamic reruns, and this move of the formatting to `TestResult` is a direct result of that since if we didn't have to change the postprocessing logic the duration formatting on line 718 would make sense.

It seems like a big PR but all I did was move a function from one place to another and it's call from one place to another so it's 2 lines. The method is the same 

@BeyondEvil

**Comment by BeyondEvil** [comment]
> It seems like a big PR but all I did was move a function from one place to another and it's call from one place to another so it's 2 lines. The method is the same
> 
> @BeyondEvil

Fair enough. 😊

**Review by BeyondEvil — APPROVED** [review]
Send it! 🚀

**Comment by gnikonorov** [comment]
Thanks @BeyondEvil! I'll merge in tomorrow assuming @ssbarnea has no issues with it.

Then I'll release `v3.1.0` on 12/2

**Pull request merged**
